server/server/src/billing/billing_service.rs

```diff
@@ -379,113 +514,300 @@ pub async fn stripe_webhooks(
+    let customer_id = match invoice
+        .customer
+        .as_ref()
+        .ok_or_else(|| {
+            ClientError(StripeWebhookError::InvalidBody(
+                "Cannot retrieve the customer_id.".to_string(),
+            ))
+        })?
+        .deref()
+    {
+        stripe::Expandable::Id(id) => id.to_string(),
+        stripe::Expandable::Object(customer) => customer.id.to_string(),
+    };
+
     let public_key: PublicKey = con
-        .maybe_json_get(public_key_from_stripe_customer_id(customer_id))
+        .maybe_json_get(public_key_from_stripe_customer_id(&customer_id))
         .await?
         .ok_or_else(|| {
             internal!("There is no public_key related to this customer_id: {:?}", customer_id)
         })?;
 
-    let user_info: StripeUserInfo = con
-        .maybe_json_get(stripe_user_info(&public_key))
-        .await?
-        .ok_or_else(|| {
-            internal!(
-                "Payment failed for a customer we don't have info about on redis: {:?}",
-                event
+    Ok(public_key)
+}
+
+#[derive(Debug)]
+pub enum GooglePlayWebhookError {
+    InvalidToken,
+    CannotRetrieveData,
+    CannotDecodePubSubData(DecodeError),
+    CannotRetrieveUserInfo,
+    CannotRetrievePublicKey,
+    CannotParseTime,
+}
+
+pub async fn google_play_notification_webhooks(
+    server_state: &Arc<ServerState>, request_body: Bytes, query_parameters: HashMap<String, String>,
+) -> Result<(), ServerError<GooglePlayWebhookError>> {
+    if !constant_time_eq::constant_time_eq(
+        query_parameters
+            .get("token")
+            .ok_or(ClientError(GooglePlayWebhookError::InvalidToken))?
+            .as_bytes(),
+        server_state.config.google.pubsub_token.as_bytes(),
+    ) {
+        return Err(ClientError(GooglePlayWebhookError::InvalidToken));
+    }
+
+    info!("Parsing pubsub notification and extracting the developer notification.");
+
+    let pubsub_notif = serde_json::from_slice::<PubSubNotification>(&request_body)?;
+    let data = base64::decode(pubsub_notif.message.data)
+        .map_err(|e| ClientError(GooglePlayWebhookError::CannotDecodePubSubData(e)))?;
+
+    let notification = serde_json::from_slice::<DeveloperNotification>(&data)?;
+
+    let mut con = server_state.index_db_pool.get().await?;
+
+    if let Some(sub_notif) = notification.subscription_notification {
+        info!("Notification is for a subscription: {:?}", sub_notif);
+
+        let subscription = google_play_client::get_subscription(
+            &server_state.android_publisher,
+            &sub_notif.subscription_id,
+            &sub_notif.purchase_token,
+        )
+        .await
+        .map_err(|e| internal!("{:#?}", e))?;
+
+        let notification_type = sub_notif.notification_type();
+        if let NotificationType::SubscriptionPurchased = notification_type {
+            return Ok(());
+        }
+
+        let account_id = &subscription
+            .obfuscated_external_account_id
```

Comment:
can you explain what part of the stack generates this id and what input it uses to do so?

Reply:
The `account_id` links a Google Play account to a Lookbook one. The `account_id` is generated on android and passed when the subscription flow first starts. Once the user has been billed, we have to confirm a subscription has been successfully added for the user (`upgrade_account_android`), we pass this generated id to the server. The server stores this id and uses it to figure out what pub/sub relates to what user.